### PR TITLE
feat(resolutions): add support for ignoring via ignore-resolutions

### PR DESCRIPTION
**Pull request by padamczewski**
### :pencil: Description

**Why?**

Currently it's not possible to conditionally handle the contents of `resolutions` field. When the property is present in `package.json` the rule will always fail. I believe this could be improved by allowing for certain packages to be ignored from that validation. A common scenario for this cases is raising hotfixes due to issues with 3rd party dependencies. I'd like to have a bit more fine-grained control on specifying exclusions when needed.

**What?**
This PR introduces `ignore-resolutions` config options which is based on the existing [ignore-packages](https://github.com/ExpediaGroup/package-json-validator/blob/456e4fd3e91c4fe71cd5baff4ae23154af426f32/src/utils/get-dependencies.ts#L24)

```
- uses: ExpediaGroup/package-json-validator@v1
    with:
      rules: resolutions
      ignore-resolutions: resolution-package-to-ignore
```

**How?**

- The default behaviour of `resolutions` doesn't change when `ignore-resolutions` is not defined: [[code](https://github.com/ExpediaGroup/package-json-validator/blob/456e4fd3e91c4fe71cd5baff4ae23154af426f32/src/utils/get-dependencies.ts#L24)], [[tests](https://github.com/ExpediaGroup/package-json-validator/compare/main...padamczewski:feat/add-resolution-ignoring-support?expand=1#diff-49ed8884a2919bb5e343d051d469a516c291824a4d09b5ec9632a668bad16565R7-R24)]

- When `ignore-resolutions` list is provided and it matches the contents of `resolutions`, the rule will pass. It will fail when there is a mis-match: [[code](https://github.com/ExpediaGroup/package-json-validator/compare/main...padamczewski:feat/add-resolution-ignoring-support?expand=1#diff-caff04b6d1929705e4fc97dd7c2f7e36da7829ee3d6a9c61f5166557d293c7beR31-R32)], [[passing-test](https://github.com/ExpediaGroup/package-json-validator/compare/main...padamczewski:feat/add-resolution-ignoring-support?expand=1#diff-49ed8884a2919bb5e343d051d469a516c291824a4d09b5ec9632a668bad16565R29-R41)], [[failing-test](https://github.com/ExpediaGroup/package-json-validator/compare/main...padamczewski:feat/add-resolution-ignoring-support?expand=1#diff-49ed8884a2919bb5e343d051d469a516c291824a4d09b5ec9632a668bad16565R48-R61)]

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -80,6 +80,19 @@ steps:
     rules: resolutions
 ```
 
+Specify `ignore-packages` to skip resolution validation entirely for certain packages. Optionally provide a newline separated list of package names here.
```
Comment (danadajian) [inline README.md:83]:
```suggestion
Specify `ignore-resolutions` to skip resolution validation entirely for certain packages. Optionally provide a newline separated list of package names here.
```

File: src/rules/resolutions.ts [hunk 2]
```diff
@@ -15,9 +15,24 @@ import * as core from '@actions/core';
 import { PackageJson } from 'type-fest';
 
 export const validateResolutions = (packageJson: PackageJson) => {
-  if (packageJson.resolutions) {
+  const ignoredResolutions = core.getMultilineInput('ignore-resolutions');
+
+  const skipIgnoreResolutions =
+    !Array.isArray(ignoredResolutions) || ignoredResolutions.length === 0;
+  if (packageJson.resolutions && skipIgnoreResolutions) {
```
Comment (danadajian) [inline src/rules/resolutions.ts:22]: `getMultilineInput()` already returns `string[]`
```suggestion
  const ignoredResolutions = core.getMultilineInput('ignore-resolutions');
  if (packageJson.resolutions && !ignoredResolutions.length) {
```

File: src/rules/resolutions.ts [hunk 3]
```diff
@@ -15,9 +15,24 @@ import * as core from '@actions/core';
...
     core.setFailed(
       'Resolutions may not be set. Please investigate the root cause of your dependency issues!'
     );
   }
+
+  if (packageJson.resolutions && Array.isArray(ignoredResolutions)) {
+    const resolutions = Object.keys(packageJson.resolutions);
+
+    const isMatching = resolutions.every(resolution => ignoredResolutions.includes(resolution));
+    if (!isMatching) {
```
Comment (danadajian) [inline src/rules/resolutions.ts:32]:
```suggestion
    const allResolutionsAreIgnored = resolutions.every(resolution => ignoredResolutions.includes(resolution));
    if (!allResolutionsAreIgnored) {
```

**Review by danadajian — COMMENTED** [review]
Thanks for contributing! Left some comments. Could you also add this new input to `action.yml`?

**Comment by padamczewski** [comment]
> Thanks for contributing!

Thanks for creating super useful action! 🙌 

> Left some comments. Could you also add this new input to `action.yml`?

Done ✅

File: README.md (hunk 1 above)
Comment (padamczewski) [reply README.md]: Silly copy paste. Thanks for picking up 👍

File: src/rules/resolutions.ts (hunk 2 above)
Comment (padamczewski) [reply src/rules/resolutions.ts]: Very true. I actually had the same thought, but didn't set up the default mock for the `@actions/core` method. 

Updated code and test to match the conditional👌

File: src/rules/resolutions.ts (hunk 3 above)
Comment (padamczewski) [reply src/rules/resolutions.ts]: Cheers, wasn't happy with the name either 🙃

File: src/rules/resolutions.ts [hunk 4]
```diff
@@ -15,9 +15,24 @@ import * as core from '@actions/core';
...
-  if (packageJson.resolutions) {
+  const ignoredResolutions = core.getMultilineInput('ignore-resolutions');
+
+  if (packageJson.resolutions && !ignoredResolutions.length) {
     core.setFailed(
       'Resolutions may not be set. Please investigate the root cause of your dependency issues!'
     );
   }
+
+  if (packageJson.resolutions && Array.isArray(ignoredResolutions)) {
```
Comment (danadajian) [inline src/rules/resolutions.ts:26]: Nit pick, but we already know `ignoredResolutions` is an array

**Review by danadajian — APPROVED** [review]

File: src/rules/resolutions.ts (hunk 4 above)
Comment (padamczewski) [reply src/rules/resolutions.ts:26]: Yes, this is another leftover. Will amend 👍 
I actually already changes this on my other branch with additional changes.

**Comment by padamczewski** [comment]
@danadajian: Wasn't sure if it fits within scope of this PR, but I figured it would be neat to allow for ignoring the failures for some period of time. Example that comes to mind:
- I notice issue with a package today, raise a PR knowing the desire is to get rid of that specific resolution
- I'd like to set until when the ignoring of `resolutions` can last
- The actions starts throwing when the set date is passed

I have a commit with the change here:
https://github.com/padamczewski/package-json-validator/commit/d38517b921ba7babe3ec07607ea76aa3874d223c 

Happy to bring it over into this PR or raise a follow-up one as well. This is assuming the change looks all right and interesting 🤞

**Review by danadajian — DISMISSED** [review]

**Comment by danadajian** [comment]
@padamczewski Let's address this in a separate PR. Definitely sounds interesting!

**Comment by padamczewski** [comment]
@danadajian: will raise another PR then👌 Also updated the `dist` folder, the checks should pass now.

**Review by danadajian — APPROVED** [review]

**Pull request merged**